rp2040-hal/src/pwm/mod.rs

```diff
@@ -0,0 +1,581 @@
+    /// Disable slice
+    #[inline]
+    pub fn disable(&mut self) {
+        self.regs.enable(true)
+    }
+
+    #[doc = "Enables phase correct mode"]
+    #[inline]
+    pub fn set_ph_correct(&mut self) {
+        self.regs.write_ph_correct(true)
+    }
+
+    #[doc = "Disales phase correct mode"]
+    #[inline]
+    pub fn clr_ph_correct(&mut self) {
+        self.regs.write_ph_correct(false)
+    }
+
+    #[doc = "Sets the integer part of the clock divider"]
+    #[inline]
+    pub fn set_div_int(&mut self, value: u8) {
+        self.regs.write_div_int(value)
+    }
+
+    #[doc = "Sets the fractional part of the clock divider"]
+    #[inline]
+    pub fn set_div_frac(&mut self, value: u8) {
+        self.regs.write_div_frac(value)
+    }
+
+    #[doc = "Sets the top register value"]
+    #[inline]
+    pub fn set_top(&mut self, value: u16) {
+        self.regs.write_top(value)
+    }
+}
+
+/// A Channel from the Pwm subsystem.
+///
+/// Its attached to one of the eight slices and can be an A or B side channel
+pub struct Channel<S: SliceId, M: SliceMode, C: ChannelId> {
+    regs: Registers<S>,
+    slice_mode: PhantomData<M>,
+    channel_id: PhantomData<C>,
+    duty_cycle: u16,
+}
+
+impl<S: SliceId, M: SliceMode, C: ChannelId> Channel<S, M, C> {
+    pub(super) unsafe fn new(duty_cycle: u16) -> Self {
+        Channel {
+            regs: Registers::new(),
+            slice_mode: PhantomData,
+            channel_id: PhantomData,
+            duty_cycle,
+        }
+    }
+}
+
+impl<S: SliceId, M: SliceMode, C: ChannelId> Sealed for Channel<S, M, C> {}
+
+impl<S: SliceId, M: SliceMode, C: ChannelId> PwmPin for Channel<S, M, C> {
+    type Duty = u16;
+
+    /// We cant disable the channel without disturbing the other channel.
+    /// So this just sets the duty cycle to zero
+    fn disable(&mut self) {
+        match C::DYN {
+            DynChannelId::A => {
+                self.duty_cycle = self.regs.read_cc_a();
+                self.regs.write_cc_a(0)
+            }
+
+            DynChannelId::B => {
+                self.duty_cycle = self.regs.read_cc_b();
+                self.regs.write_cc_b(0)
+            }
+        }
+    }
+
+    fn enable(&mut self) {
+        match C::DYN {
+            DynChannelId::A => self.regs.write_cc_a(self.duty_cycle),
+
+            DynChannelId::B => self.regs.write_cc_b(self.duty_cycle),
+        }
+    }
+
+    fn get_duty(&self) -> Self::Duty {
+        match C::DYN {
+            DynChannelId::A => self.regs.read_cc_a(),
+            DynChannelId::B => self.regs.read_cc_b(),
+        }
+    }
+
+    fn get_max_duty(&self) -> Self::Duty {
+        self.regs.read_top()
+    }
+
+    fn set_duty(&mut self, duty: Self::Duty) {
+        match C::DYN {
+            DynChannelId::A => self.regs.write_cc_a(duty),
+            DynChannelId::B => self.regs.write_cc_b(duty),
+        }
+    }
+}
+
+impl<S: SliceId, M: SliceMode + ValidSliceMode<S>> Channel<S, M, A> {
+    ///
+    pub fn output_to<
```

Comment:
This should probably return `PwmPinToken` because otherwise there is no way to recover the pin back from PWM.